### PR TITLE
highlight moused primitives

### PR DESCRIPTION
**Pull request by ShiboSoftwareDev**

### CODE REVIEW

File: src/components/CanvasElementsRenderer.tsx [hunk 1]
```diff
@@ -27,13 +28,36 @@ export interface CanvasElementsRendererProps {
 }
 
 export const CanvasElementsRenderer = (props: CanvasElementsRendererProps) => {
+  const getGlobalId = useGlobalStore((state) => state.getGlobalId)
```
Comment (seveibar) [inline src/components/CanvasElementsRenderer.tsx:31]: Instead of putting this on the global store, just define it as a mutating global method:

```
let globalIdCounter = 0
function getGlobalId() {
  return `globalid_${globalIdCounter++}`
}
```

Comment (seveibar) [reply src/components/CanvasElementsRenderer.tsx:31]: this may fix the console issue

File: src/components/CanvasElementsRenderer.tsx [hunk 2]
```diff
@@ -27,13 +28,36 @@ export interface CanvasElementsRendererProps {
...
+    const traces: AnySoupElement[] = []
+    const primitives = props.elements.flatMap((elm) => {
+      if (elm.type === "source_trace") traces.push(elm)
+      return convertElementToPrimitives(elm, props.elements, getGlobalId)
+    })
+
+    return [primitives, traces]
   }, [props.elements])
+
+  const highlightedConnectedPrimitives = useMemo(() => {
```
Comment (seveibar) [inline src/components/CanvasElementsRenderer.tsx:47]:
```suggestion
  const primitivesWithHighlights = useMemo(() => {
```

File: src/components/CanvasElementsRenderer.tsx [hunk 3]
```diff
@@ -27,13 +28,36 @@ export interface CanvasElementsRendererProps {
...
-  const primitives = useMemo(() => {
-    const primitives = props.elements.flatMap((elm) =>
-      convertElementToPrimitives(elm, props.elements),
-    )
-    return primitives
+  const [primitives, traces] = useMemo(() => {
+    const traces: AnySoupElement[] = []
+    const primitives = props.elements.flatMap((elm) => {
+      if (elm.type === "source_trace") traces.push(elm)
+      return convertElementToPrimitives(elm, props.elements, getGlobalId)
```
Comment (seveibar) [inline src/components/CanvasElementsRenderer.tsx:41]:
```suggestion
      return convertElementToPrimitives(elm, props.elements)
```

File: src/components/MouseElementTracker.tsx [hunk 4]
```diff
@@ -53,8 +75,10 @@ export const MouseElementTracker = ({
         same_space_index,
       })
     }
+    if (!containSameIds(newHighlightedPrimitives, highlightedPrimitives))
+      setHighlightedPrimitives(newHighlightedPrimitives)
 
-    return highlightedPrimitives
+    return newHighlightedPrimitives
```
Comment (seveibar) [inline src/components/MouseElementTracker.tsx:81]: you should undo most of the changes in this file and just change line 115 so that it's 

```ts
setMousedPrimitives(mousedPrimitives)
setHighlightedPrimitives(mousedPrimitives.map(p => p._drawing_object_id)
```

Comment (ShiboSoftwareDev) [reply src/components/MouseElementTracker.tsx:81]: I did the check so that highlightedPrimitives don't change on every mouse movement, it updated so frequently. Is that okay?

File: src/lib/getElectricallyConnectedPrimitiveIds.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { AnySoupElement } from "@tscircuit/soup"
+import { Primitive } from "./types"
+import { HighlightedPrimitive } from "components/MouseElementTracker"
+
+export const getElectricallyConnectedPrimitiveIds = (
```
Comment (seveibar) [inline src/lib/getElectricallyConnectedPrimitiveIds.ts:5]: Not a huge fan of this function definition, very hard to follow.

```ts
isElementElectricallyConnected(elm: AnySoupElement, allElements: AnySoupElement[])
```

that signature makes a lot of sense and can be used to determine if an element is highlighted

```ts
if (isElementElectricallyConnected(primitive._element, soup)) {
  primitive.highlighted = true
}
```

File: src/lib/Drawer.ts [hunk 6]
```diff
@@ -349,6 +352,11 @@ export class Drawer {
       ctx.fillStyle = "rgba(0,0,0,1)"
       ctx.strokeStyle = "rgba(0,0,0,1)"
     }
+
+    if (highlighted) {
+      ctx.fillStyle = lighten(ctx.fillStyle, 0.5)
+      ctx.strokeStyle = lighten(ctx.strokeStyle, 0.5)
+    }
```
Comment (seveibar) [inline src/lib/Drawer.ts:359]: nice

File: src/components/MouseElementTracker.tsx (hunk 4 above)
Comment (seveibar) [reply src/components/MouseElementTracker.tsx:81]: yea that's fine, it also applied to moused primitives. Let's give it a better name though

Probably something like this, also have to add support for the connected primitives


```tsx

if (hasSamePrimitives(oldMousedPrimitives, newMousedPrimitives)) return

setMousedPrimitives(mousedPrimitives)
setHighlightedPrimitives(
  mousedPrimitives.flatMap(mousedPrimitive => getAllConnectedPrimitives({
    mousedPrimitive,
    allPrimitives: primitives,
  }))
)
```

**Review by seveibar — CHANGES_REQUESTED** [review]
looking a lot better. We can also break this PR into two parts:

1. Highlighting only moused primitives
2. Highlighting any primitive connected to a moused primitive

When I worked on universal data tool, i let a lot of PRs in that IMO didn't have super clear code and it eventually made the project really difficult to contribute to, so that's why i'm being a bit more difficult on this PR, because I know the consequences of a UI project if it starts to get really complicated. Would definitely recommend breaking into two parts!!!

**Comment by ShiboSoftwareDev** [comment]
@seveibar I'll do my best

**Comment by seveibar** [comment]
Looking pretty good!!

**Comment by seveibar** [comment]
One issue is we shouldnt have a big object inside the global store if we can avoid it, this is why i suggest primitive ids as a list

**Comment by seveibar** [comment]
I'm going to try to do a quick video review again as soon as i get home from a conference!
